### PR TITLE
Module config as a map

### PR DESCRIPTION
**Pull request by chrzaszcz**
Start the conversion of module options from proplists to maps. Maps are more efficient, easier to work with, and better integrated with `mongoose_config`.

Changes in `gen_mod`:
- Support for maps in option lookup, including nested maps. Proplists need to be supported at least until all modules are converted to use maps.
- More strict types for option keys and values.
  - This caused some rework of a few places in the code, where values were functions. It is not possible to set an arbitrary `fun` in TOML, so I think it's best to avoid this.
  - Instead of setting functions, set module names which implement a specific behaviour, like we do for backends.
  - For now, `merger_fun` in `mongoose_user_cache` is hardcoded to be `maps:merge/2` as it was impossible to set this option anyway.
- Dropped support for `set_opt` to discourage modifying options at runtime - if we need to do this with maps, the built-in map syntax is better suited for this. The result should be a new type.
- Removed calls to `gen_mod` API not related to modules, e.g. general proplist processing, but only where `dialyzer` complained.

Changes in modules:
- Update two first modules, `mod_auth_token` and `mod_extdisco`, to use maps.
- Defaults are set up in the config specs.
- `mod_extdisco` supports dynamic domains now - it was easy and the code was already modified a lot.
- I decided to use direct pattern matching on maps to get the opts in simple cases instead of calling `gen_mod:get_opt/2`. I am not completely sure if this is always the right way, but it makes sense to leverage the language constructs instead of calling getter functions. However, the getters for nested maps might be useful, so I think that we need to find out what works best as we progress with the remaining modules.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104939) / small_tests / f968609566ed5b88b855d8fd5e244fd83712b1a0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3534/104939/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3534/104939/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-07_07.29.50/lib.mongooseim.logs/run.2022-02-07_07.29.51/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/105008) / small_tests / 3e286cf099f780b286adbb9d6022b5a1be471e70
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3534/105008/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3534/105008/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-07_12.49.55/lib.mongooseim.logs/run.2022-02-07_12.49.56/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/105044) / small_tests / 20876d2cd9bb50775495021eb66c8386b458bdee
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3534/105044/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3534/105044/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-07_13.36.46/lib.mongooseim.logs/run.2022-02-07_13.36.46/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/105062) / small_tests / dcb75df13e3887e54fe6e96d423c1f8c9479b149
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3534/105062/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3534/105062/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-07_14.35.41/lib.mongooseim.logs/run.2022-02-07_14.35.41/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_23](https://circleci.com/gh/esl/MongooseIM/105079) / small_tests / 4bb0778754152d0ed2329a55d0643af9aea06c5b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3534/105079/small_tests.23.1) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3534/105079/small_tests.23.1/small/ct_run.mongooseim@localhost.2022-02-07_14.49.16/lib.mongooseim.logs/run.2022-02-07_14.49.16/suite.log.html)

File: src/mod_extdisco.erl [hunk 1]
```diff
@@ -20,66 +20,70 @@
...
-                                        validate = port},
-                  <<"transport">> => #option{type = string,
-                                             validate = {enum, ["udp", "tcp"]}},
-                  <<"username">> => #option{type = string,
-                                            validate = non_empty},
-                  <<"password">> => #option{type = string,
-                                            validate = non_empty}
-            },
-        required = [<<"type">>, <<"host">>]
+       items = #{<<"type">> => #option{type = atom,
```
Comment (gustawlippa) [inline src/mod_extdisco.erl:56]: Minor, but I like the old indentation more.

**Review by gustawlippa — APPROVED** [review]
Nice, looks good! 👍

File: src/mod_extdisco.erl (hunk 1 above)
Comment (chrzaszcz) [reply src/mod_extdisco.erl]: That's Erlang Mode for Emacs in action... I changed the syntax to avoid this strange case.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/105266) / small_tests / 503b2497e53eb57f3031c5926a17e73b81315418
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3534/105266/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3534/105266/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-09_16.03.13/lib.mongooseim.logs/run.2022-02-09_16.03.14/suite.log.html)

**Pull request merged**